src/bitcoin/electrum/mod.rs

```diff
@@ -198,25 +200,42 @@ impl Electrum {
         }
-        let changeset = self.bdk_wallet.apply_connected_chain_update(chain_update);
-
-        let mut changes_iter = changeset.into_iter();
-        let reorg_common_ancestor = loop {
-            match changes_iter.next() {
-                Some((height, Some(_))) => {
-                    // `BlockHash` being `Some(_)` means a checkpoint at this height was added to the chain.
-                    // Since we iterate in ascending height order, we'll see the lowest block height first.
-                    // If the lowest height it adds is higher than our height before syncing, we're good.
-                    // Else if it's adding a block at height before syncing or lower, it's a reorg.
-                    break if height > local_chain_tip.height() {
-                        None
-                    } else {
-                        log::info!("Block chain reorganization detected.");
-                        Some(self.bdk_wallet.find_block_at_or_before_height(height))
-                    };
+
+        let reorg_common_ancestor = if !is_rescanning {
+            let changeset = self.bdk_wallet.apply_connected_chain_update(chain_update);
+            let mut changes_iter = changeset.into_iter();
+            loop {
+                match changes_iter.next() {
+                    Some((height, Some(_))) => {
+                        // `BlockHash` being `Some(_)` means a checkpoint at this height was added to the chain.
+                        // Since we iterate in ascending height order, we'll see the lowest block height first.
+                        // If the lowest height it adds is higher than our height before syncing, we're good.
+                        // Else if it's adding a block at height before syncing or lower, it's a reorg.
+                        break if height > local_chain_tip.height() {
+                            None
+                        } else {
+                            log::info!("Block chain reorganization detected.");
+                            Some(self.bdk_wallet.find_block_at_or_before_height(height))
+                        };
+                    }
+                    Some((_, None)) => continue,
+                    None => break None,
                 }
-                Some((_, None)) => continue,
-                None => break None,
+            }
+        } else {
+            // In case of a full scan, just replace our local chain with the new one.
+            self.bdk_wallet.replace_chain_from_tip(chain_update);
+            // In case our chain tip was previously positive, treat this full scan
+            // as a reorg with common ancestor being the genesis block.
+            if local_chain_tip.height() > 0 {
+                Some(tip_from_block_id(
+                    self.local_chain()
+                        .get(0)
+                        .expect("contains genesis")
+                        .block_id(),
+                ))
```

Comment:
I don't understand the rationale for this. What should we unconditionally treat rescans as reorgs?

Reply:
This was mainly to keep it simple and avoid adding logic to check for a reorg and get common ancestor (assuming we can't rely on the chain update for that). Given that we'll roll back the DB to genesis anyway as part of rescan logic, I felt the safest thing was just to do that also here.